### PR TITLE
[paved path] basic charnn with GPT example

### PR DESCRIPTION
**Pull request by hudeven**
1) We decide to put paved path examples in torchrecipes repo
2) This is an initial commit based on previous work https://github.com/aivanou/disttraining and https://github.com/dracifer/disttraining
3) It's a basic example without any external dependencies(mlflow logging, airflow, torchx, etc), which will be added soon

Testing:

* single GPU
`python main.py`
* multi GPUs
`torchrun --nnodes 1 --nproc_per_node 4 \
--rdzv_backend c10d \
--rdzv_endpoint localhost:29500 \
main.py`

* torch single GPU
`torchx run  -s local_cwd dist.ddp -j 1x1 --script main.py`

* torchx multi GPUs
`torchx run  -s local_cwd dist.ddp -j 1x4 --script main.py`

### CODE REVIEW

File: torchrecipes/paved_path/charnn/model.py [hunk 1]
```diff
@@ -0,0 +1,194 @@
+#!/usr/bin/env python3
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+"""
+The model taken form https://github.com/karpathy/minGPT
```
Comment (edward-io) [inline torchrecipes/paved_path/charnn/model.py:9]: from*

File: torchrecipes/paved_path/charnn/model.py [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+    n_layer: int
+    n_head: int
+    n_embd: int
+    embd_pdrop: float = 0.1
+    resid_pdrop: float = 0.1
+    attn_pdrop: float = 0.1
+
+
+class EmbeddingStem(nn.Module):
+    def __init__(self, config: GPTConfig, device="cpu", dtype=torch.float32):
```
Comment (edward-io) [inline torchrecipes/paved_path/charnn/model.py:66]: type for device

**Review by edward-io — COMMENTED** [review]
is this ready for review? sorry i started adding comments before asking

File: torchrecipes/paved_path/charnn/main.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+        dist.init_process_group("gloo", rank=rank, world_size=world_size)
+
+
+def generate_seq(cfg: DictConfig, model: torch.nn.Module, dataset: CharDataset) -> None:
+    if dist.get_rank() == 0:
+        device = get_device()
+        context = cfg["charnn"]["phrase"]
+        x = torch.tensor([dataset.stoi[s] for s in context], dtype=torch.long)[
+            None, ...
+        ].to(device)
```
[the remark this answers is not in the document]
Comment (hudeven) [reply torchrecipes/paved_path/charnn/main.py:106]: Yes, they are equivalent. The former is used more in numpy. unsqueeze(0) looks more "pytorchic" though

File: torchrecipes/paved_path/charnn/main.py [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+        "TORCHELASTIC_RUN_ID", str(uuid.uuid4()).split("-")[0]
+    )
+
+
+def get_job_name():
+    uid = os.environ["TORCHELASTIC_RUN_ID"]
+    return f"test-job-{uid}"
+
+
+def get_device() -> Optional[int]:
```
[the remark this answers is not in the document]
Comment (hudeven) [reply torchrecipes/paved_path/charnn/main.py]: good idea! I made the change

**Comment by hudeven** [comment]
@edward-io yes, thanks for taking a look! As it's the initial check in from previous prototyping repos, there are a lot of things to improve! Any feedback is welcome!

File: torchrecipes/paved_path/charnn/main.py [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+
+
+def get_device() -> torch.device:
+    if  torch.cuda.is_available():
+        local_rank = int(os.environ.get("LOCAL_RANK", "0"))
+        device = torch.device(f"cuda:{local_rank}")
+        torch.cuda.set_device(device)
+    else:
+        device = torch.device("cpu")
+    return device
```
Comment (ananthsub) [inline torchrecipes/paved_path/charnn/main.py:56]: should we use the util here? 

https://github.com/pytorch/tnt/blob/5fd77ae9ad8357079cebd14532cb40a2ce8e5761/torchtnt/utils/device.py#L20-L48

File: torchrecipes/paved_path/charnn/main.py [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+        model = model.to(device)
+        device_ids = [device.index]
+    model = DistributedDataParallel(
+        model,
+        device_ids=device_ids,
+    )
+    optimizer = torch.optim.AdamW(
+        model.parameters(), lr=opt_config.lr, weight_decay=opt_config.weight_decay
+    )
+    return model, optimizer
```
Comment (ananthsub) [inline torchrecipes/paved_path/charnn/main.py:80]: should we try out torchsnapshot here? @yifuwang ?

File: torchrecipes/paved_path/charnn/main.py [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+
+
+def setup_process_group() -> None:
+    device = get_device()
+    rank = int(os.environ["RANK"])
+    world_size = int(os.environ["WORLD_SIZE"])
+    if device.type == "cuda":
+        dist.init_process_group("nccl", rank=rank, world_size=world_size)
+    else:
+        dist.init_process_group("gloo", rank=rank, world_size=world_size)
```
Comment (ananthsub) [inline torchrecipes/paved_path/charnn/main.py:102]: assuming we're launching with torchrun/torchx, this would initialize both device & process group: 

https://github.com/pytorch/tnt/blob/5fd77ae9ad8357079cebd14532cb40a2ce8e5761/torchtnt/utils/env.py#L34-L38

File: torchrecipes/paved_path/charnn/main.py [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+
+
+def generate_seq(cfg: DictConfig, model: torch.nn.Module, dataset: CharDataset) -> None:
+    if dist.get_rank() == 0:
```
Comment (ananthsub) [inline torchrecipes/paved_path/charnn/main.py:106]: is distributed always initialized? 

we cover this here: https://github.com/pytorch/tnt/blob/5fd77ae9ad8357079cebd14532cb40a2ce8e5761/torchtnt/utils/distributed.py#L98-L110

File: torchrecipes/paved_path/charnn/main.py [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+        context = cfg["charnn"]["phrase"]
+        x = torch.tensor([dataset.stoi[s] for s in context], dtype=torch.long, device=device).unsqueeze(0)
+        y = sample(model, x, 2000, temperature=1.0, sample=True, top_k=10)[0]
+        completion = "".join([dataset.itos[int(i)] for i in y])
+        print(completion)
+
+
+def set_seed(seed: int) -> None:
+    random.seed(seed)
+    torch.manual_seed(seed)
```
Comment (ananthsub) [inline torchrecipes/paved_path/charnn/main.py:117]: https://github.com/pytorch/tnt/blob/5fd77ae9ad8357079cebd14532cb40a2ce8e5761/torchtnt/utils/seed.py#L19-L38

**Comment by hudeven** [comment]
@ananthsub thanks for the comments! All of them make sense. I plan to adopt those libs soon. Currently, the main goal of this pull request is to add the basic example and then have pipeline setup config/instructions on top of it.
